File: go_engine/engine/metrics.go
package engine

import "sync"

type Counter struct {
	values map[string]float64
	mu     sync.Mutex
}

func (c *Counter) Inc(label string, v float64) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.values[label] = c.values[label] + v
}

func (c *Counter) Snapshot() map[string]float64 {
	c.mu.Lock()
	defer c.mu.Unlock()
	out := make(map[string]float64, len(c.values))
	for k, v := range c.values {
		out[k] = v
	}
	return out
}

type Histogram struct {
	counts map[string]int
	sums   map[string]float64
	mu     sync.Mutex
}

func (h *Histogram) Observe(label string, v float64) {
	h.mu.Lock()
	defer h.mu.Unlock()
	h.counts[label] = h.counts[label] + 1
	h.sums[label] = h.sums[label] + v
}

func (h *Histogram) Snapshot() map[string]map[string]float64 {
	h.mu.Lock()
	defer h.mu.Unlock()
	out := make(map[string]map[string]float64, len(h.counts))
	for k, count := range h.counts {
		out[k] = map[string]float64{
			"count": float64(count),
			"sum":   h.sums[k],
		}
	}
	return out
}

type MetricsRegistry struct {
	counters   map[string]*Counter
	histograms map[string]*Histogram
	mu         sync.Mutex
}

func NewMetricsRegistry() *MetricsRegistry {
	return &MetricsRegistry{
		counters:   map[string]*Counter{},
		histograms: map[string]*Histogram{},
	}
}

func (m *MetricsRegistry) Counter(name string) *Counter {
	m.mu.Lock()
	defer m.mu.Unlock()
	if _, ok := m.counters[name]; !ok {
		m.counters[name] = &Counter{values: map[string]float64{}}
	}
	return m.counters[name]
}

func (m *MetricsRegistry) Histogram(name string) *Histogram {
	m.mu.Lock()
	defer m.mu.Unlock()
	if _, ok := m.histograms[name]; !ok {
		m.histograms[name] = &Histogram{
			counts: map[string]int{},
			sums:   map[string]float64{},
		}
	}
	return m.histograms[name]
}

func (m *MetricsRegistry) Snapshot() map[string]interface{} {
	m.mu.Lock()
	defer m.mu.Unlock()
	out := map[string]interface{}{}
	counterSnapshot := map[string]map[string]float64{}
	for name, counter := range m.counters {
		counterSnapshot[name] = counter.Snapshot()
	}
	histSnapshot := map[string]map[string]map[string]float64{}
	for name, hist := range m.histograms {
		histSnapshot[name] = hist.Snapshot()
	}
	out["counters"] = counterSnapshot
	out["histograms"] = histSnapshot
	return out
}

File: python_engine/workflow_engine/metrics.py
"""
Simple in-memory metrics registry.
"""

from __future__ import annotations

import threading
import time
from dataclasses import dataclass
from typing import Dict, Tuple


LabelKey = Tuple[Tuple[str, str], ...]


def _normalize_labels(labels: Dict[str, str] | None) -> LabelKey:
    if not labels:
        return ()
    return tuple(sorted((str(k), str(v)) for k, v in labels.items()))


@dataclass
class Counter:
    name: str
    _values: Dict[LabelKey, float]
    _lock: threading.Lock

    def inc(self, value: float = 1.0, labels: Dict[str, str] | None = None) -> None:
        key = _normalize_labels(labels)
        with self._lock:
            self._values[key] = self._values.get(key, 0.0) + value

    def snapshot(self) -> Dict[LabelKey, float]:
        with self._lock:
            return dict(self._values)


@dataclass
class Histogram:
    name: str
    _counts: Dict[LabelKey, int]
    _sums: Dict[LabelKey, float]
    _lock: threading.Lock

    def observe(self, value: float, labels: Dict[str, str] | None = None) -> None:
        key = _normalize_labels(labels)
        with self._lock:
            self._counts[key] = self._counts.get(key, 0) + 1
            self._sums[key] = self._sums.get(key, 0.0) + value

    def snapshot(self) -> Dict[LabelKey, Dict[str, float]]:
        with self._lock:
            return {
                key: {"count": float(self._counts.get(key, 0)), "sum": self._sums.get(key, 0.0)}
                for key in set(self._counts) | set(self._sums)
            }


class MetricsRegistry:
    def __init__(self) -> None:
        self._counters: Dict[str, Counter] = {}
        self._histograms: Dict[str, Histogram] = {}
        self._lock = threading.Lock()

    def counter(self, name: str) -> Counter:
        with self._lock:
            if name not in self._counters:
                self._counters[name] = Counter(name=name, _values={}, _lock=threading.Lock())
            return self._counters[name]

    def histogram(self, name: str) -> Histogram:
        with self._lock:
            if name not in self._histograms:
                self._histograms[name] = Histogram(name=name, _counts={}, _sums={}, _lock=threading.Lock())
            return self._histograms[name]

    def timer(self, name: str, labels: Dict[str, str] | None = None):
        histogram = self.histogram(name)

        class _Timer:
            def __enter__(self_inner):
                self_inner.start = time.time()
                return self_inner

            def __exit__(self_inner, exc_type, exc, tb):
                duration = time.time() - self_inner.start
                histogram.observe(duration, labels=labels)
                return False

        return _Timer()

    def snapshot(self) -> Dict[str, Dict]:
        return {
            "counters": {name: counter.snapshot() for name, counter in self._counters.items()},
            "histograms": {name: hist.snapshot() for name, hist in self._histograms.items()},
        }
